adr/0016-hooks.md

```diff
@@ -0,0 +1,113 @@
+
+Currently there are 2 solutions:
+
+1. The package deployer solves the problem pre-deployment (creating needed repos, secrets, etc...)
+2. The package itself solves these problems with `actions` that are custom built for ECR clusters.
+
+Neither one of these current solutions are ideal. We don't want to require overly complex external + prior actions for Zarf package deployments, and we don't want package creators to have to create and distribute packages that are specific to ECR.
+
+## Approaches Considered
+
+### Internal Zarf Implementation
+ Clusters that have hooks will have `zarf-hook-*` secret(s) in the 'zarf' namespace. This secret will contain the hook's configuration and any other required metadata. As part of the package deployment process, Zarf will check if the cluster has any hooks and run them if they exist. Given the scenario above, there is no longer a need for an ECR specific Zarf package to be created. An ECR hook would perform the proper configuration for any package deployed onto that cluster; thereby requiring no extra manual intervention from the package deployer.
+
+
+Zarf HookConfig state information struct:
+```go
+type HookConfig struct {
+	HookName     string                 `json:"hookName" jsonschema:"description=Name of the hook"`
+	Internal     bool                   `json:"internal" jsonschema:"description=Internal hooks are run by Zarf itself, not by a plugin"`
+	Lifecycle    HookLifecycle          `json:"lifecycle" jsonschema:"description=Lifecycle of the hook"`
+	HookData     map[string]interface{} `json:"hookData" jsonschema:"description=Generic data map used for the hook. The data is obtained from a secret in the Zarf namespace"`
+	OCIReference string                 `json:"ociReference" jsonschema:"description=Optional OCI reference to the hook image to run"`
+}
+```
+
+Example Secret Data:
+```yaml
+hookName: ecr-repository
+internal: true
+lifecycle: before-component
+hookData:
+  registryURL: public.ecr.aws/abcdefg/zarf-ecr-registry
+  region: us-east-1
+  repositoryPrefix: ecr-zarf-registry
+```
+
+For this solution, hooks have to be 'installed' onto a cluster before they are used. When Zarf is deploying a package onto a cluster, it will look for any secrets with the `zarf-hook` label in the `zarf` namespace.  If hooks are found, Zarf will run any 'package' level hooks before deploying a component and run any 'component' level hook for each component that is getting deployed. The hook lifecycle options will be before/after a package deployment and before/after a component deployment.
+ - NOTE: The order of hook execution is nearly random. If there are multiple hooks for a lifecycle there is no guarantee that they will be executed in a certain order.
+ - NOTE: The `package` lifecycle might be changed to a `run-once` lifecycle. This would benefit packages that don't have kube context information when the deployment starts.
```

Comment:
We should probably say what all of the lifecycle events we expect to be are (I don't see a list of them).

Reply:
The new parent to that note is:
> Hooks lifecycle options will be before/after a package deployment and before/after a component deployment.

Does that satisfy what you'd like to see for expected lifecycle events?